Firestore/core/test/firebase/firestore/model/field_path_test.cc

```diff
@@ -0,0 +1,272 @@
+TEST(FieldPath, Comparison) {
+  const FieldPath abc{"a", "b", "c"};
+  const FieldPath abc2{"a", "b", "c"};
+  const FieldPath xyz{"x", "y", "z"};
+  EXPECT_EQ(abc, abc2);
+  EXPECT_NE(abc, xyz);
+
+  const FieldPath empty;
+  const FieldPath a{"a"};
+  const FieldPath b{"b"};
+  const FieldPath ab{"a", "b"};
+
+  EXPECT_TRUE(empty < a);
+  EXPECT_TRUE(a < b);
+  EXPECT_TRUE(a < ab);
+
+  EXPECT_TRUE(a > empty);
+  EXPECT_TRUE(b > a);
+  EXPECT_TRUE(ab > a);
+}
+
+TEST(FieldPath, IsPrefixOf) {
+  const FieldPath empty;
+  const FieldPath a{"a"};
+  const FieldPath ab{"a", "b"};
+  const FieldPath abc{"a", "b", "c"};
+  const FieldPath b{"b"};
+  const FieldPath ba{"b", "a"};
+
+  EXPECT_TRUE(empty.IsPrefixOf(empty));
+  EXPECT_TRUE(empty.IsPrefixOf(a));
+  EXPECT_TRUE(empty.IsPrefixOf(ab));
+  EXPECT_TRUE(empty.IsPrefixOf(abc));
+  EXPECT_TRUE(empty.IsPrefixOf(b));
+  EXPECT_TRUE(empty.IsPrefixOf(ba));
+
+  EXPECT_FALSE(a.IsPrefixOf(empty));
+  EXPECT_TRUE(a.IsPrefixOf(a));
+  EXPECT_TRUE(a.IsPrefixOf(ab));
+  EXPECT_TRUE(a.IsPrefixOf(abc));
+  EXPECT_FALSE(a.IsPrefixOf(b));
+  EXPECT_FALSE(a.IsPrefixOf(ba));
+
+  EXPECT_FALSE(ab.IsPrefixOf(empty));
+  EXPECT_FALSE(ab.IsPrefixOf(a));
+  EXPECT_TRUE(ab.IsPrefixOf(ab));
+  EXPECT_TRUE(ab.IsPrefixOf(abc));
+  EXPECT_FALSE(ab.IsPrefixOf(b));
+  EXPECT_FALSE(ab.IsPrefixOf(ba));
+
+  EXPECT_FALSE(abc.IsPrefixOf(empty));
+  EXPECT_FALSE(abc.IsPrefixOf(a));
+  EXPECT_FALSE(abc.IsPrefixOf(ab));
+  EXPECT_TRUE(abc.IsPrefixOf(abc));
+  EXPECT_FALSE(abc.IsPrefixOf(b));
+  EXPECT_FALSE(abc.IsPrefixOf(ba));
+}
+
+TEST(FieldPath, AccessFailures) {
+  const FieldPath path;
+  ASSERT_DEATH_IF_SUPPORTED(path.front(), "");
+  ASSERT_DEATH_IF_SUPPORTED(path.back(), "");
+  ASSERT_DEATH_IF_SUPPORTED(path[0], "");
+  ASSERT_DEATH_IF_SUPPORTED(path[1], "");
+  ASSERT_DEATH_IF_SUPPORTED(path.at(0), "");
+  ASSERT_DEATH_IF_SUPPORTED(path.DropFirst(), "");
+  ASSERT_DEATH_IF_SUPPORTED(path.DropFirst(2), "");
+  ASSERT_DEATH_IF_SUPPORTED(path.DropLast(), "");
+}
+
+TEST(FieldPath, Parsing) {
+  const auto parse = [](const std::pair<std::string, size_t> expected) {
+    const auto path = FieldPath::ParseServerFormat(expected.first);
+    return std::make_pair(path.CanonicalString(), path.size());
+  };
+  const auto make_expected = [](const std::string& str, const size_t size) {
+    return std::make_pair(str, size);
+  };
+
+  auto expected = make_expected("foo", 1);
+  EXPECT_EQ(expected, parse(expected));
+  expected = make_expected("foo.bar", 2);
+  EXPECT_EQ(expected, parse(expected));
+  expected = make_expected("foo.bar.baz", 3);
+  EXPECT_EQ(expected, parse(expected));
+  expected = make_expected(R"(`.foo\\`)", 1);
+  EXPECT_EQ(expected, parse(expected));
+  expected = make_expected(R"(`.foo\\`.`.foo`)", 2);
+  EXPECT_EQ(expected, parse(expected));
+  expected = make_expected(R"(foo.`\``.bar)", 3);
+  EXPECT_EQ(expected, parse(expected));
+
+  const auto path_with_dot = FieldPath::ParseServerFormat(R"(foo\.bar)");
```

Comment:
I added an equivalent test to Objective C. At first I thought it's a bug, but now I look at it, I guess it's reasonable.